src/test/java/io/vertx/test/core/HttpConnectionEarlyResetTest.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2011-2016 The original author or authors
+ * ------------------------------------------------------
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Apache License v2.0 which accompanies this distribution.
+ *
+ *     The Eclipse Public License is available at
+ *     http://www.eclipse.org/legal/epl-v10.html
+ *
+ *     The Apache License v2.0 is available at
+ *     http://www.opensource.org/licenses/apache2.0.php
+ *
+ * You may elect to redistribute this code under either of these licenses.
+ */
+
+package io.vertx.test.core;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.vertx.core.http.HttpServer;
+import io.vertx.core.http.HttpServerOptions;
+import io.vertx.core.http.impl.HttpServerImpl;
+import io.vertx.core.impl.VertxInternal;
+import io.vertx.core.net.NetClientOptions;
+import io.vertx.core.net.NetSocket;
+import org.junit.Test;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Make sure that the Netty pipeline has a handler catching the {@link java.io.IOException} if the connection is reset
+ * before any data has been sent.
+ *
+ * @author Thomas Segismont
+ */
+public class HttpConnectionEarlyResetTest extends VertxTestBase {
+
+  private HttpServer httpServer;
+  private AtomicReference<Throwable> caught = new AtomicReference<>();
+  private CountDownLatch resetLatch = new CountDownLatch(1);
+
+  @Override
+  public void setUp() throws Exception {
+    super.setUp();
+    CountDownLatch listenLatch = new CountDownLatch(1);
+    httpServer = new HttpServerImpl((VertxInternal) vertx, new HttpServerOptions()) {
+      @Override
+      protected void onChannelInitialized(Channel ch) {
+        ch.pipeline().addFirst(new ResetLatchCountDown()).addLast(new ThrowableRecorder());
```

Comment:
Just pushed another version. Tell me if this is what you had in mind.

2016-09-13 14:03 GMT+02:00 Julien Viet <notifications@github.com>:

> In src/test/java/io/vertx/test/core/HttpConnectionEarlyResetTest.java
> <https://github.com/eclipse/vert.x/pull/1620#discussion_r78544578>:
>
> > + * @author Thomas Segismont
> > + */
> > +public class HttpConnectionEarlyResetTest extends VertxTestBase {
> > +
> > +  private HttpServer httpServer;
> > +  private AtomicReference<Throwable> caught = new AtomicReference<>();
> > +  private CountDownLatch resetLatch = new CountDownLatch(1);
> > +
> > +  @Override
> > +  public void setUp() throws Exception {
> > +    super.setUp();
> > +    CountDownLatch listenLatch = new CountDownLatch(1);
> > +    httpServer = new HttpServerImpl((VertxInternal) vertx, new HttpServerOptions()) {
> > +      @Override
> > +      protected void onChannelInitialized(Channel ch) {
> > +        ch.pipeline().addFirst(new ResetLatchCountDown()).addLast(new ThrowableRecorder());
>
> I would rather be in favor of adding a connectionExceptionHandler in
> HttpServerImpl (not exposed) that would be called by the pipeline, the
> default implementation closes the channel.
>
> The test can then change this handler to be aware of the failure and test
> properly.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/vert.x/pull/1620/files/bd2ef07fc8e7c9060bf8db0b5c38360094a25ad7#r78544578>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABblttlap9c5k1X9it7M7HwNDGN64gWDks5qppEWgaJpZM4J44vk>
> .
>